### ISSUE TITLE
landmark-unique flagging elements inside of iframe

### ISSUE THREAD

**Issue by krikienoid** [dequelabs/axe-core#2592]
I have a web page with a `<header role="banner">` landmark, and there is also an `<iframe>` on the page which has it's own banner landmark as well. Axe is flagging it with the "Ensures landmarks are unique" rule, even though these elements belong to different DOM trees. This same issue is also happening with contentinfo landmarks.

I thought the solution would be to add `role="document"` to the iframe but that did not make a difference. According to this [W3C article](https://www.w3.org/TR/wai-aria-practices/examples/landmarks/banner.html): 

> When a page contains nested document and/or application roles (e.g. typically through the use of iframe and frame elements), each document or application role may have one banner landmark.

So my question is, is Axe giving a false negative, or is this the expected behavior? If this is expected, can the [documentation](https://dequeuniversity.com/rules/axe/4.0/landmark-unique?application=AxeChrome) be updated with a passing example of an iframe that has its own landmarks?

axe-core version: 4.0.2
axe Accessibility Extension version: 4.6.1

**Comment by straker** [comment]
Thanks for the issue. @WilcoFiers would be a better judge of if this is intended behavior or not.

**Comment by WilcoFiers** [comment]
This is intentional, and where axe-core is in disagreement with ARIA documentation. ARIA has this idea of document and applications that can be nested, where you can put documents in documents in applications in documents. Assistive technologies don't really support this. The only one that actual does is VoiceOver. In Safari with VO, an iframe is something that you enter into, in the same way that the web page itself is a document you either navigate past or enter into. No other screen reader (that I know of) behaves like that. A page is just one document, iframe content included.

I haven't tested this recently, but my recollection is that if you navigate between landmarks on a page, the landmarks in iframes are treated no differently than those in the in the top level page (again, except for VoiceOver). Axe-core's approach is more akin to how WCAG treats web pages. The content of an iframe or an object element in WCAG isn't treated as its own page, but as part of the page it is loaded into.

In other words, while yes, the ARIA specs says documents can be nested and each nested document can have its own landmarks, VoiceOver is the only screen reader that actually follows this model. That is also why axe-core tests heading levels aren't skipped across frame boundaries. To most screen readers, what is and isn't in a frame makes no difference, so axe-core doesn't make that distinction either.

**Comment by krikienoid** [comment]
Thank you for the detailed explanation. So would you recommend that developers should avoid having iframe content that contains its own landmarks then? It doesn't sound like there's really any other way to approach this.

**Comment by WilcoFiers** [comment]
Yeah, except maybe if the iframe contains most of the page. But that almost never happens.

**Comment by bcalou** [comment]
Hello,
4 years later, do you think that assistive technologies evolved on this, @WilcoFiers ?
